src/main/java/org/wildfly/naming/client/AbstractContext.java

```diff
@@ -67,6 +70,28 @@ protected AbstractContext() {
     }
 
     /**
+     * Construct a new instance.  A new, empty environment map is created for this context.
+     *
+     * @param namingProviders the naming providers to use, may be {@code null}
+     */
+    protected AbstractContext(final NamingProvider[] namingProviders) {
```

Comment:
Thanks for the feedback, David! Keeping AbstractContext provider-agnostic does make sense and I agree that having a single NamingProvider in NamingContextFactory implementations does simplify things. I really like the idea of moving the selection logic to the provider instead. I'll work on these changes.